File: app/backend/peaks.py
"""Named summits from OpenStreetMap via the Overpass API (best effort: failures return no labels)."""
import json
import math
import os
import re
import time
import urllib.error
import urllib.parse
import urllib.request

from cache import SharedCache

OVERPASS_URL = os.environ.get("OVERPASS_URL", "https://overpass-api.de/api/interpreter")
NOMINATIM_URL = os.environ.get("NOMINATIM_URL", "https://nominatim.openstreetmap.org/search")
USER_AGENT = "ridgeline-studio/0.1 (prototype)"
PEAK_TYPES = {"peak", "volcano"}  # OSM natural= values worth offering as a "mountain" search result

_cache = SharedCache(maxsize=512)  # (bbox, limit) -> summits; failed lookups raise and aren't cached
_search_cache = SharedCache(maxsize=256)  # (query, limit) -> search results


def parse_ele(value):
    """OSM `ele` tags are messy: "3798", "3798 m", "3.798" (German thousands separator)."""
    match = re.match(r"\s*(-?\d+)(?:[.,](\d+))?", value or "")
    if not match:
        return None
    whole, fraction = match.groups()
    if fraction and len(fraction) == 3 and abs(int(whole)) < 10:
        return int(whole + fraction)
    return round(float(f"{whole}.{fraction or 0}"))


def _query_overpass(request, attempts=2):
    """POST a query; the public instance often answers 429/5xx when busy and succeeds right after."""
    for attempt in range(attempts):
        try:
            # Can take ~8 s even for a small frame; summits never block the lines, so wait generously
            with urllib.request.urlopen(request, timeout=30) as response:
                return json.load(response)["elements"]
        except urllib.error.HTTPError as e:
            if e.code not in (429, 502, 503, 504) or attempt == attempts - 1:
                raise
            time.sleep(1)


def summits(bbox, limit=6):
    """The highest named peaks in bbox, spread out so labels don't pile up. Returns None if the lookup fails."""
    try:
        return _cache.get((bbox, limit), lambda: _lookup(bbox, limit))
    except Exception:
        return None


def _lookup(bbox, limit):
    west, south, east, north = bbox
    query = f'[out:json][timeout:25];node["natural"="peak"]["name"]["ele"]({south},{west},{north},{east});out body;'
    request = urllib.request.Request(
        OVERPASS_URL,
        data=urllib.parse.urlencode({"data": query}).encode(),
        headers={"User-Agent": USER_AGENT},
    )
    elements = _query_overpass(request)

    candidates = []
    for el in elements:
        ele = parse_ele(el["tags"].get("ele"))
        if ele is not None:
            candidates.append({"name": el["tags"]["name"], "ele": ele, "lat": el["lat"], "lon": el["lon"]})
    candidates.sort(key=lambda p: -p["ele"])

    # Greedy: highest first, skipping summits closer than 15% of the frame diagonal to one already chosen
    cos_lat = math.cos(math.radians((south + north) / 2))
    min_distance = 0.15 * math.hypot((east - west) * cos_lat, north - south)
    chosen = []
    for p in candidates:
        if all(math.hypot((p["lon"] - q["lon"]) * cos_lat, p["lat"] - q["lat"]) >= min_distance for q in chosen):
            chosen.append(p)
            if len(chosen) == limit:
                break

    return tuple(chosen)


def search_peaks(query, limit=8):
    """Free-text mountain search via Nominatim (OSM's geocoder), for the search bar's autocomplete.

    Returns None if the lookup fails so the caller can distinguish "no results" from "unavailable".
    """
    query = query.strip()
    if not query:
        return ()
    try:
        return _search_cache.get((query.lower(), limit), lambda: _search(query, limit))
    except Exception:
        return None


def _search(query, limit):
    params = {
        "q": query,
        "format": "jsonv2",
        "addressdetails": 1,
        "extratags": 1,
        "namedetails": 1,
        "accept-language": "en",  # otherwise country/address names come back in the local language
        "limit": 20,  # over-fetch: most matches for a free-text query aren't peaks, filtered out below
    }
    request = urllib.request.Request(
        f"{NOMINATIM_URL}?{urllib.parse.urlencode(params)}",
        headers={"User-Agent": USER_AGENT},
    )
    with urllib.request.urlopen(request, timeout=10) as response:
        elements = json.load(response)

    results = []
    for el in elements:
        # jsonv2 calls the OSM primary tag "category"; older Nominatim responses call it "class"
        if el.get("category", el.get("class")) != "natural" or el.get("type") not in PEAK_TYPES:
            continue
        name = (el.get("namedetails") or {}).get("name") or el["display_name"].split(",")[0]
        results.append({
            "name": name,
            "lat": float(el["lat"]),
            "lon": float(el["lon"]),
            "elevation": parse_ele((el.get("extratags") or {}).get("ele")),
            "country": (el.get("address") or {}).get("country"),
        })
        if len(results) == limit:
            break
    return tuple(results)

File: app/server/utils/peaks.ts
/**
 * Named summits from OpenStreetMap via the Overpass API, and free-text mountain search via
 * Nominatim. Port of app/backend/peaks.py — best effort: failures return null, not a thrown
 * error, so the route can decide how to degrade (empty labels, "unavailable" message, ...).
 */
import type { Bbox, PeakSearchResult, RawSummit } from '../../types/ridgeline'
import { SharedCache } from './sharedCache'

const USER_AGENT = 'skycomb-studio/0.1 (prototype)'
const PEAK_TYPES = new Set(['peak', 'volcano']) // OSM natural= values worth offering as a "mountain" search result

const summitsCache = new SharedCache<string, RawSummit[]>(512)
const searchCache = new SharedCache<string, PeakSearchResult[]>(256)

/** OSM `ele` tags are messy: "3798", "3798 m", "3.798" (German thousands separator). */
export function parseEle(value: string | null | undefined): number | null {
  if (!value) return null
  const match = /^\s*(-?\d+)(?:[.,](\d+))?/.exec(value)
  if (!match) return null
  const [, whole, fraction] = match
  if (fraction && fraction.length === 3 && Math.abs(Number(whole)) < 10) {
    return Number(whole + fraction)
  }
  return Math.round(Number(`${whole}.${fraction ?? 0}`))
}

function bboxKey(bbox: Bbox): string {
  return bbox.map((v) => v.toFixed(4)).join(',')
}

/** POST a query; the public instance often answers 429/5xx when busy and succeeds right after. */
async function queryOverpass(query: string, attempts = 2): Promise<any[]> {
  const config = useRuntimeConfig()
  for (let attempt = 0; attempt < attempts; attempt++) {
    // Can take ~8s even for a small frame; summits never block the lines, so wait generously
    const response = await fetch(config.overpassUrl, {
      method: 'POST',
      headers: { 'User-Agent': USER_AGENT, 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({ data: query }),
      signal: AbortSignal.timeout(30_000),
    })
    if (response.ok) return (await response.json()).elements
    if (![429, 502, 503, 504].includes(response.status) || attempt === attempts - 1) {
      throw new Error(`Overpass request failed: HTTP ${response.status}`)
    }
    await new Promise((resolve) => setTimeout(resolve, 1000))
  }
  throw new Error('Overpass request failed')
}

/** The highest named peaks in bbox, spread out so labels don't pile up. Null if the lookup fails. */
export async function summits(bbox: Bbox, limit = 6): Promise<RawSummit[] | null> {
  try {
    return await summitsCache.get(`${bboxKey(bbox)}|${limit}`, () => lookupSummits(bbox, limit))
  } catch {
    return null
  }
}

async function lookupSummits(bbox: Bbox, limit: number): Promise<RawSummit[]> {
  const [west, south, east, north] = bbox
  const query = `[out:json][timeout:25];node["natural"="peak"]["name"]["ele"](${south},${west},${north},${east});out body;`
  const elements = await queryOverpass(query)

  const candidates: RawSummit[] = []
  for (const el of elements) {
    const ele = parseEle(el.tags?.ele)
    if (ele !== null) candidates.push({ name: el.tags.name, ele, lat: el.lat, lon: el.lon })
  }
  candidates.sort((a, b) => b.ele - a.ele)

  // Greedy: highest first, skipping summits closer than 15% of the frame diagonal to one already chosen
  const cosLat = Math.cos((((south + north) / 2) * Math.PI) / 180)
  const minDistance = 0.15 * Math.hypot((east - west) * cosLat, north - south)
  const chosen: RawSummit[] = []
  for (const p of candidates) {
    const farEnough = chosen.every((q) => Math.hypot((p.lon - q.lon) * cosLat, p.lat - q.lat) >= minDistance)
    if (farEnough) {
      chosen.push(p)
      if (chosen.length === limit) break
    }
  }
  return chosen
}

/**
 * Free-text mountain search via Nominatim (OSM's geocoder), for the search bar's autocomplete.
 * Returns null if the lookup fails so the caller can distinguish "no results" from "unavailable".
 */
export async function searchPeaks(query: string, limit = 8): Promise<PeakSearchResult[] | null> {
  const trimmed = query.trim()
  if (!trimmed) return []
  try {
    return await searchCache.get(`${trimmed.toLowerCase()}|${limit}`, () => search(trimmed, limit))
  } catch {
    return null
  }
}

async function search(query: string, limit: number): Promise<PeakSearchResult[]> {
  const config = useRuntimeConfig()
  const params = new URLSearchParams({
    q: query,
    format: 'jsonv2',
    addressdetails: '1',
    extratags: '1',
    namedetails: '1',
    'accept-language': 'en', // otherwise country/address names come back in the local language
    limit: '20', // over-fetch: most matches for a free-text query aren't peaks, filtered out below
  })
  const response = await fetch(`${config.nominatimUrl}?${params}`, {
    headers: { 'User-Agent': USER_AGENT },
    signal: AbortSignal.timeout(10_000),
  })
  if (!response.ok) throw new Error(`Nominatim request failed: HTTP ${response.status}`)
  const elements = (await response.json()) as any[]

  const results: PeakSearchResult[] = []
  for (const el of elements) {
    // jsonv2 calls the OSM primary tag "category"; older Nominatim responses call it "class"
    const category = el.category ?? el.class
    if (category !== 'natural' || !PEAK_TYPES.has(el.type)) continue
    const name = el.namedetails?.name || el.display_name.split(',')[0]
    results.push({
      name,
      lat: Number(el.lat),
      lon: Number(el.lon),
      elevation: parseEle(el.extratags?.ele),
      country: el.address?.country ?? null,
    })
    if (results.length === limit) break
  }
  return results
}
